### PR TITLE
Support Flax networks in `wandb.watch`

### LINKED ISSUE #11263: [Feature]: Add support for Flax networks in `wandb.watch`
### Description

Hi! First of all thank you for this great project, it’s been a pleasure to use so far.

I noticed that `wandb.watch` does not currently support Flax models. Given the growing adoption of Flax/JAX workflows, I believe native support would be pretty useful for tracking parameters and gradients in Flax networks, similarly to what’s already available for PyTorch.

### Suggested Solution

I'd be happy to work on this trying to mimic the same logic used with the PyTorch integration, if that sounds reasonable. Looking forward to your feedback!

### PR DESCRIPTION
**Pull request by flferretti**
Description
-----------

- Fixes #11263 

What does the PR do? Include a concise description of the PR contents.

This PR adds support for Flax networks in `wandb.watch` in a similar manner as it has been done for PyTorch integration.

- [x] I updated CHANGELOG.unreleased.md, or it's not applicable

Testing
-------
How was this PR tested?

A test file has been added in order to test the main functionalities, in the same fashion of the pytorch integration test

### CODE REVIEW

File: wandb/integration/flax/wandb_flax.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+                value, grads = original_vg_fn(params, *fn_args, **fn_kwargs)
+                # Capture for logging
+                global _captured_grads, _captured_params
+                _captured_grads = grads
+                _captured_params = params
+                return value, grads
+
+            return value_and_grad_fn
+
+        # Monkey-patch JAX
```
Comment (AakashKumarNain) [inline wandb/integration/flax/wandb_flax.py:169]: The monkey patching here is bad. I am not sure it will even work with jit.

File: wandb/integration/flax/wandb_flax.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        self._original_value_and_grad = jax.value_and_grad
+
+        # Replace with wrapped versions
+        def wrapped_grad(fun: Callable, *args, **kwargs) -> Callable:
+            original_grad_fn = self._original_grad(fun, *args, **kwargs)
+
+            def grad_fn(params, *fn_args, **fn_kwargs):
+                grads = original_grad_fn(params, *fn_args, **fn_kwargs)
+                # Capture for logging
+                global _captured_grads, _captured_params
```
Comment (AakashKumarNain) [inline wandb/integration/flax/wandb_flax.py:149]: Isn't that a side effect?

**Comment by AakashKumarNain** [comment]
Thanks for the PR. I am unsure if this is the right way to watch a JAX model like this. I can see certain pattern here that won't play well. One thing you can do to test it out is to build an actual model with more than one layer, and run it with jitted patterns to notice the side effects. 

PS: I believe that for JAX the philosophy of logging should be "you log what you run but explicitly". It is the only way to ensure that you are not running silent side-effects.

**Comment by flferretti** [comment]
> Thanks for the PR. I am unsure if this is the right way to watch a JAX model like this. I can see certain pattern here that won't play well. One thing you can do to test it out is to build an actual model with more than one layer, and run it with jitted patterns to notice the side effects.
> 
> PS: I believe that for JAX the philosophy of logging should be "you log what you run but explicitly". It is the only way to ensure that you are not running silent side-effects.

Hi @AakashKumarNain, thanks a lot for the review and for calling out the issues with the current approach. 

After digging into it more, I agree that monkey-patching JAX is not a sound solution for Flax. A potentially better alternative may be to adopt a `watch_step`-style API rather than trying to auto-capture gradients implicitly:

```python
run.watch(model, log="all", log_freq=100)

@run.watch_step(returns=("state", "loss", "grads"))
@jax.jit
def train_step(state, batch):
    ...
    return state, loss, grads
```

In that design, the actual train step stays fully JIT-compiled and the wrapper runs outside JIT. The wrapper can inspect the returned state / grads and log parameter + gradient histograms according to the existing watch config and log_freq, without introducing side effects inside the compiled function. That would not be perfect PyTorch-style parity, but it seems much more compatible with JAX while still giving users a reasonably ergonomic watch workflow.

If this direction makes sense, I’m happy to rework the PR around `watch_step` instead of the current approach. Otherwise, feel free to close this without merging
